keps/0035-json-schema-export.md

```diff
@@ -0,0 +1,296 @@
+owners:
+  - "@aneumann82"
+creation-date: 2020-09-22
+last-updated: 2020-09-29
+status: provisional
+see-also:
+  - KEP-33
+---
+
+# JSON-schema export
+
+## Table of Contents
+
+<!--ts-->
+   * [JSON-schema export](#json-schema-export)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+            * [Story 1](#story-1)
+         * [Implementation Details](#implementation-details)
+            * [Additional Types:](#additional-types)
+            * [Additional Attributes](#additional-attributes)
+            * [Groups](#groups)
+         * [JSON-Schema export](#json-schema-export-1)
+            * [JSON-Schema output](#json-schema-output)
+      * [Alternatives](#alternatives)
+      * [Implementation History](#implementation-history)
+
+<!-- Added by: aneumann, at: Thu Oct  1 09:47:25 CEST 2020 -->
+
+<!--te-->
+
+## Summary
+
+Extend the existing `params.yaml` with metadata and new types and add an export as JSON-schema.
+
+## Motivation
+
+JSON-schema is the preferred way to describe datastructures for parameters and generate user-friendly UIs. KUDO should be able to export a JSON-schema that describes the required parameters.
+
+KEP-33 proposes a full replacement of the existing flat list `params.yaml` with a nested and typed structure described by JSON-schema. This is a big breaking change and requires a lot of work to get right in a single release.
+
+This KEP proposes to enhance the flat list parameters with additional attributes and add support to export a JSON-schema. This is a first step towards KEP-33, it allows tooling to be build based on JSON-Schema while not introducing breaking changes on the operator level.
+
+### Goals
+
+- Add additional attributes required for a useful JSON-schema and UI generation
+- Add support to export a JSON-schema that describes the parameters
+- Prepare Implementation of KEP-33 in the future 
+
+### Non-Goals
+
+- Handle multiple CRD versions
+- Replace the `params.yaml` with a json-schema
+- Add validation for parameters
+- Change the format how KUDO reads parameters
+
+## Proposal
+
+Add new types as well as optional new attributes to enhance the parameters and bring them closer to JSON-Schema.
+
+Implement an export that generates a valid JSON-Schema out of the flat parameter list.
+
+### User Stories
+
+#### Story 1
+
+A user wants to generate a UI to install a KUDO operator. He wants to use an automatic generator, for example https://rjsf-team.github.io/react-jsonschema-form/
+
+### Implementation Details
+
+#### Additional Types: 
+
+To bring the parameters up to par with JSON-schema, we add these additional types:
+- `boolean` A "true" or "false" value, from the JSON "true" or "false" productions
+- `number` An arbitrary-precision, base-10 decimal number value, from the JSON "number" production
```

Comment:
In Javascript and therefore in JSON, `number` is a double-precision float, should we use that definition here instead?